container-search/src/main/java/com/yahoo/search/dispatch/rpc/RpcProtobufFillInvoker.java

```diff
@@ -135,34 +141,59 @@ private void sendDocsumsRequest(int nodeId, List<FastHit> hits, byte[] payload,
             log.warning("Got hits with node id " + nodeId + ", which is not included in the current dispatch config");
             return;
         }
-
         Query query = result.getQuery();
         Compressor.Compression compressionResult = compressor.compress(query, payload);
         node.request(RPC_METHOD, compressionResult.type(), payload.length, compressionResult.data(),
                 roe -> receive(roe, hits), clientTimeout);
     }
 
+    private ResponseAndHits getNextResponse(Result result) throws InterruptedException {
+        long timeLeftMs = result.getQuery().getTimeLeft();
+        if (timeLeftMs <= 0) {
+            return null;
+        }
+        var responseAndHits = responses.poll(timeLeftMs, TimeUnit.MILLISECONDS);
+        if (responseAndHits == null || responseAndHits.response().timeout()) {
+            return null;
+        }
+        return responseAndHits;
+    }
+
     private void processResponses(Result result, String summaryClass) throws TimeoutException {
         try {
-            int skippedHits = 0;
+            List<FastHit> skippedHits = new ArrayList<>();
             while (outstandingResponses > 0) {
-                long timeLeftMs = result.getQuery().getTimeLeft();
-                if (timeLeftMs <= 0) {
-                    throwTimeout();
-                }
-                var responseAndHits = responses.poll(timeLeftMs, TimeUnit.MILLISECONDS);
+                var responseAndHits = getNextResponse(result);
                 if (responseAndHits == null) {
                     throwTimeout();
                 }
-                var response = responseAndHits.getFirst();
-                if (response.timeout()) {
-                    throwTimeout();
-                }
-                var hitsContext = responseAndHits.getSecond();
-                skippedHits += processResponse(result, response, hitsContext, summaryClass);
+                skippedHits.addAll(processOneResponse(result, responseAndHits, summaryClass, false));
                 outstandingResponses--;
             }
-            if (skippedHits != 0) {
+            if (! skippedHits.isEmpty()) {
+                ListMap<Integer, FastHit> retryMap = new ListMap<>();
+                for (Integer nodeId : resourcePool.knownNodeIds()) {
+                    for (var hit : skippedHits) {
+                        if (hit.getDistributionKey() != nodeId) {
+                            retryMap.put(nodeId, hit);
+                        }
+                    }
+                }
+                if (retryMap.size() > 0) {
+                    sendFillRequestByNode(result, summaryClass, retryMap);
+                }
+                while (outstandingResponses > 0) {
+                    var responseAndHits = getNextResponse(result);
+                    if (responseAndHits == null) {
+                        log.log(Level.WARNING, "Timed out waiting for summary data. " + outstandingResponses + " responses outstanding.");
```

Comment:
Do we want to log as 'warning' or could 'fine' suffice? Could a container overload scenario be worsen by excessive logging?